pest/src/prec_climber.rs

```diff
@@ -79,137 +91,322 @@ impl<R: RuleType> BitOr for Operator<R> {
-    /// # enum Rule {
-    /// #     plus,
-    /// #     minus,
-    /// #     times,
-    /// #     divide,
-    /// #     power
-    /// # }
-    /// PrecClimber::new(vec![
-    ///     Operator::new(Rule::plus, Assoc::Left) | Operator::new(Rule::minus, Assoc::Left),
-    ///     Operator::new(Rule::times, Assoc::Left) | Operator::new(Rule::divide, Assoc::Left),
-    ///     Operator::new(Rule::power, Assoc::Right)
-    /// ]);
-    /// ```
-    pub fn new(ops: Vec<Operator<R>>) -> PrecClimber<R> {
-        let ops = ops.into_iter()
-            .zip(1..)
-            .fold(HashMap::new(), |mut map, (op, prec)| {
-                let mut next = Some(op);
-
-                while let Some(op) = next.take() {
-                    match op {
-                        Operator {
-                            rule,
-                            assoc,
-                            next: op_next
-                        } => {
-                            map.insert(rule, (prec, assoc));
-                            next = op_next.map(|op| *op);
-                        }
-                    }
-                }
 
-                map
-            });
-
-        PrecClimber { ops }
-    }
-
-    /// Performs the precedence climbing algorithm on the `pairs` in a similar manner to map-reduce.
-    /// *Primary* pairs are mapped with `primary` and then reduced to one single result with
-    /// `infix`.
-    ///
-    /// # Panics
-    ///
-    /// Panics will occur when `pairs` is empty or when the alternating *primary*, *operator*,
-    /// *primary* order is not respected.
-    ///
-    /// # Examples
-    ///
-    /// ```ignore
-    /// let primary = |pair| {
-    ///     consume(pair, climber)
-    /// };
-    /// let infix = |lhs: i32, op: Pair<Rule>, rhs: i32| {
-    ///     match op.rule() {
-    ///         Rule::plus => lhs + rhs,
-    ///         Rule::minus => lhs - rhs,
-    ///         Rule::times => lhs * rhs,
-    ///         Rule::divide => lhs / rhs,
-    ///         Rule::power => lhs.pow(rhs as u32),
-    ///         _ => unreachable!()
-    ///     }
-    /// };
-    ///
-    /// let result = climber.climb(pairs, primary, infix);
-    /// ```
-    pub fn climb<'i, P, F, G, T>(&self, mut pairs: P, mut primary: F, mut infix: G) -> T
+    /// Instantiate a new `PrecedenceClimber`.
+    pub fn new() -> Self {
+        Self {
+            prec: 1,
+            ops: HashMap::new(),
+        }
+    }
+
+    /// Add `op` to `PrecedenceClimber`.
+    pub fn op(mut self, op: Op<R>) -> Self {
+        self.prec += 1;
+        let mut iter = Some(op);
+        while let Some( Op { rule, kind, next } ) = iter.take() {
+            self.ops.insert(rule, (kind, self.prec));
+            iter = next.map(|op| *op);
+        }
+        self
+    }
+
+    /// Maps primary expressions with a closure `primary`.
+    pub fn map_primary<'climber, 'pest, X, T>(&'climber self, primary: X)
```

Comment:
`'i` in pest stands for `input`. `'pest` lifetime is too ambiguous.

```suggestion
    pub fn map_primary<'climber, 'i, X, T>(&'climber self, primary: X)
```

Reply:
done